### PR TITLE
[KernelGen][Nvidia] Add alpha_dropout operator with Triton kernel

### PR DESCRIPTION
**Pull request by XDYuanzhuLee**
## Summary
Adds a Triton kernel for `alpha_dropout`. Applies alpha dropout to the input. Alpha Dropout is a type of Dropout
that maintains the self-normalizing property by scaling activations
with SELU activation function parameters.

## Testing
- Validated against reference on device via `to_reference(inp, True)`
- Tested on: Nvidia, Tianshu, Muxi, Ascend, Hygon

## Performance
Test command: `pytest benchmark/test_alpha_dropout.py --level core` (NVIDIA H20)

### alpha_dropout

| dtype | Size | Torch Latency (ms) | Gems Latency (ms) | Speedup |
|-------|------|--------------------|-------------------|---------|
| float16 | [64, 64] | 0.019008 | 0.243152 | 0.078 |
| float16 | [1024, 1024] | 0.028672 | 0.241232 | 0.119 |
| float16 | [4096, 4096] | 0.174480 | 0.264320 | 0.660 |
| float16 | [64, 512, 512] | 0.175456 | 0.263104 | 0.667 |
| float16 | [1024, 1024, 1024] | 10.346976 | 11.630528 | 0.890 |
| float32 | [64, 64] | 0.019136 | 0.239424 | 0.080 |
| float32 | [1024, 1024] | 0.032512 | 0.240736 | 0.135 |
| float32 | [4096, 4096] | 0.288608 | 0.307776 | 0.938 |
| float32 | [64, 512, 512] | 0.290624 | 0.307552 | 0.945 |
| float32 | [1024, 1024, 1024] | 16.771168 | 17.411455 | 0.963 |
| bfloat16 | [64, 64] | 0.018176 | 0.235456 | 0.077 |
| bfloat16 | [1024, 1024] | 0.029504 | 0.234752 | 0.126 |
| bfloat16 | [4096, 4096] | 0.174496 | 0.256320 | 0.681 |
| bfloat16 | [64, 512, 512] | 0.175456 | 0.255328 | 0.687 |
| bfloat16 | [1024, 1024, 1024] | 10.256064 | 11.769024 | 0.871 |

| Operator | Arithmetic Mean Speedup |
|----------|------------------------|
| alpha_dropout | **0.528** |

## Multi-backend Testing
| Backend | Accuracy Test | Speedup (mean) | Notes |
|---|---|---|---|
| Nvidia (H20) | PASS (15 cases) | 0.528 | Primary |
| Tianshu | PASS | 6.601 | — |
| Muxi | PASS | 1.005 | — |
| Ascend | FAIL | — | if (metadata_path is None) and ( |
| Hygon | FAIL | 5.462 | Error: tuple index out of range |

## Files Changed
- `src/flag_gems/ops/alpha_dropout.py`: Triton kernel implementation
- `tests/test_alpha_dropout.py`: Accuracy test
- `benchmark/test_alpha_dropout.py`: Performance benchmark
- `src/flag_gems/ops/__init__.py`: Register import and `__all__`
- `src/flag_gems/__init__.py`: Register to `_FULL_CONFIG`
- `conf/operators.yaml`: Add operator entry (kind: NeuralNetwork, stage: alpha 5.1)

### CODE REVIEW

**Review by Dongxu-H — COMMENTED** [review]
## FlagGems Operator Contribution Review

**Operator**: `alpha_dropout`  
**PR**: #3908

### 📊 Change Statistics
| Metric | Value |
|--------|-------|
| Additions | 222 |
| Deletions | 0 |
| Files Changed | 6 |

### 📋 22-Point Checklist

| # | Requirement | Status |
|---|-------------|--------|
| 1 | Implementation file path correctness | ✅ Correct: `src/flag_gems/ops/alpha_dropout.py` |
| 2 | KernelGen attribution in header | ✅ Present |
| 3 | Import order (stdlib → third-party → internal) | ⚠️ Could not verify |
| 4 | Function signature alignment with PyTorch API | ⚠️ Needs manual verification |
| 5 | Type support documentation | ⚠️ Not checked |
| 6 | Triton kernel decorators (`@triton.jit`, etc.) | ⚠️ `@triton.jit` present, verify other decorators |
| 7 | Test file existence | ✅ Present: `tests/test_alpha_dropout.py` |
| 8 | Test coverage (shapes, dtypes) | ✅ Shape and dtype tests present |
| 9 | Benchmark file existence | ✅ Present: `benchmark/test_alpha_dropout.py` |
| 10 | Benchmark implementation correctness | ✅ Benchmark implemented |
| 11 | Config registration (`operators.yaml`) | ✅ Registered in `operators.yaml` |
| 12 | Module registration (`__init__.py`) | ✅ Registered in `__init__.py` |
| 13 | Ops module registration (`ops/__init__.py`) | ✅ Registered in `ops/__init__.py` |
| 14 | Logger namespace (`flag_gems.*`) | ⚠️ Missing `flag_gems.` prefix |
| 15 | Error handling | ✅ Error handling present |
| 16 | Numerical stability patterns | ✅ Stability patterns found |
| 17 | Memory efficiency | ⚠️ Needs profiling |
| 18 | Edge case tests | ⚠️ Needs verification |
| 19 | Backend support (Metax) | ❌ Not implemented |
| 20 | Documentation/Docstrings | ✅ Docstrings present |
| 21 | Code style consistency | ⚠️ Run linter manually |
| 22 | Integration verification | ⚠️ Verify end-to-end |

### 📈 Summary

| Category | Count |
|----------|-------|
| ✅ Pass | 12 |
| ⚠️ Warning | 9 |
| ❌ Fail | 1 |

**Overall**: ⚠️ Needs Attention

---

*This review was generated by an automated script following the FlagGems Operator Contribution Guidelines.*
*For detailed code review, please consult a domain expert.*

**Review by Dongxu-H — COMMENTED** [review]
## 💡 审查建议

建议:
- ⚠️ 建议在flag_gems/__init__.py注册

**Review by Dongxu-H — COMMENTED** [review]
## 💡 审查建议

建议:
- ⚠️ 建议在flag_gems/__init__.py注册

**Review by Dongxu-H — COMMENTED** [review]
## 💡 审查建议

建议:
- ⚠️ 建议在flag_gems/__init__.py注册

**Review by Dongxu-H — APPROVED** [review]
## Code Review for PR 3908 - alpha_dropout operator

### ✅ Approval

Well-implemented alpha dropout for SELU activation with proper affine transform to maintain self-normalizing property.

#### `src/flag_gems/ops/alpha_dropout.py`

**Strengths:**

1. **Correct SELU Constants**
   ```python
   _ALPHA = 1.6732632423543772848170429916717
   _SCALE = 1.0507009873554804934193349852946
   _ALPHA_PRIME = -_ALPHA * _SCALE  # -1.7580993408473766
   ```
   Matches the SELU paper (Klambauer et al. 2017) values exactly.

2. **Correct Affine Transform**
   ```python
   def _alpha_dropout_affine(p: float):
       a = 1.0 / math.sqrt((1.0 - p) * (1.0 + p * _ALPHA_PRIME * _ALPHA_PRIME))
       b = -a * p * _ALPHA_PRIME
       return a, b
   ```
   This preserves mean=0 and variance=1, which is the key property of alpha dropout.

3. **Proper Philox RNG Usage**
   Correctly uses Philox backend for reproducible random number generation with proper seed/offset handling.

4. **Edge Cases Handled**
   ```python
   if not train or p == 0:
       return input.clone()
   if p == 1:
       return torch.zeros_like(input)
   ```
   Good handling of eval mode and extreme p values.

5. **UNROLL Optimization**
   ```python
   UNROLL: tl.constexpr = 4
   ```
   Processing 4 elements per kernel instance for efficiency.

**Minor suggestions:**

1. **Duplicate alpha_prime Definition**
   ```python
   _ALPHA_PRIME = -_ALPHA * _SCALE  # -1.7580993408473766  # at module level
   alpha_prime = -1.7580993408473766  # in kernel
   ```
   Consider using `_ALPHA_PRIME` directly in the kernel to avoid duplication.

2. **Validation Message**
   ```python
   assert 0.0 < p < 1.0, "p must be in (0, 1)"
   ```
   Good, but PyTorch's alpha_dropout accepts p=0 and p=1 (handled above). The assertion should be `0.0 <= p <= 1.0` or remove it since edge cases are already handled.

#### `tests/test_alpha_dropout.py`

**Good test approach:**
- Tests both train=True and train=False (eval mode)
- Uses statistical validation for stochastic operation
- Handles vendor-specific seeding for kunlunxin

**Minor observation:**
The test validates shape and dtype but not statistical properties (mean≈0, variance≈1). For a complete test, consider:
```python
# Verify statistical properties (mean should be ~0, variance ~1)
assert abs(res_out.mean()) < 0.1  # loose tolerance
assert abs(res_out.std() - 1.0) < 0.2  # loose tolerance
```

### ✅ What Looks Good

- Correct SELU affine transform mathematics
- Proper Philox RNG implementation
- Edge case handling
- UNROLL optimization
- Good documentation in docstring

This PR is ready to merge. The implementation correctly follows the SELU dropout paper.

**Review by Caeruleann — APPROVED** [review]
